main.go

```diff
@@ -44,8 +56,163 @@ func usage(out io.Writer) {
 	fmt.Fprintf(out, "  -k, --no-verify      Do not verify TLS/SSL certificates.\n")
 	fmt.Fprintf(out, "  -r, --redirect       Rewrite and return 3xx redirects.\n")
 	fmt.Fprintf(out, "      --keep-referer   Pass Referer: headers through, even with -r.\n")
+	fmt.Fprintf(out, "      --tls            Present TLS (with a custom CA) to clients connecting\n")
+	fmt.Fprintf(out, "                       to us.  The CA will be dumped to standard error.\n")
+}
+
+type Cert struct {
+	RawCertificate *x509.Certificate
+	RawKey         *rsa.PrivateKey
+
+	Certificate string
+	Key         string
+}
+
+func (ca *Cert) Sign(cert *Cert) error {
+	raw, err := x509.CreateCertificate(rand.Reader, cert.RawCertificate, ca.RawCertificate, cert.RawKey.Public(), ca.RawKey)
+	if err != nil {
+		return err
+	}
+
+	cert.Certificate = string(pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: raw,
+	}))
+
+	return nil
+}
+
+func certificate(name string, serial int, ttl time.Duration) (*Cert, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	cert := &x509.Certificate{
+		SignatureAlgorithm:    x509.SHA512WithRSA, /* FIXME: hard-coded */
+		PublicKeyAlgorithm:    x509.RSA,
+		Subject:               pkix.Name{CommonName: name},
+		DNSNames:              []string{name},
+		IPAddresses:           []net.IP{net.ParseIP("127.0.0.1")},
```

Comment:
Are we worried about people (ab)using gotcha and listening on things other than loopback via the `bind` argument?  We do run gotcha as a cf app for arbitrary proxying logic, but that's all (currently) over http.

Think about it, and let me know what you decide.